src/kmeans.jl

```diff
@@ -1,91 +1,132 @@
 # K-means algorithm
 
-#### Interface
-
-mutable struct KmeansResult{T<:Real} <: ClusteringResult
-    centers::Matrix{T}         # cluster centers (d x k)
-    assignments::Vector{Int}   # assignments (n)
-    costs::Vector{T}           # costs of the resultant assignments (n)
-    counts::Vector{Int}        # number of samples assigned to each cluster (k)
-    cweights::Vector{Float64}  # cluster weights (k)
-    totalcost::Float64         # total cost (i.e. objective) (k)
-    iterations::Int            # number of elapsed iterations
-    converged::Bool            # whether the procedure converged
+####
+#### Result object
+####
+
+# NOTE REFACT: there's no need for this to be a mutable struct.
+
+struct KmeansResult{T<:AbstractFloat} <: ClusteringResult
+    centers::Matrix{T}        # cluster centers (d x k)
+    assignments::Vector{Int}  # assignments (n)
+    costs::Vector{T}          # cost of the assignments (n)
```

Comment:
I don't see why `costs` and `weights` have to be equal to `T`.
I guess, `eltype(costs)` should be something like `promote_type(typeof(float(abs2(zero(T))))), eltype(weights))`.

Reply:
there are in the PR after conversion (see `kmeans!`). But maybe wait a bit for the full review as I'm implementing further changes to remove more conversions as per comments in #140 .